src/cursor/abstract_cursor.ts

```diff
@@ -691,243 +695,156 @@ export abstract class AbstractCursor<
-
-  if (error) {
-    if (cursor.loadBalanced && error instanceof MongoNetworkError) {
-      return await completeCleanup();
+      await this.cleanup();
     }
   }
 
-  if (cursorId == null || server == null || cursorId.isZero() || cursorNs == null) {
-    if (needsToEmitClosed) {
-      cursor[kClosed] = true;
-      cursor[kId] = Long.ZERO;
-      cursor.emit(AbstractCursor.CLOSE);
-    }
-
-    if (session) {
-      if (session.owner === cursor) {
+  /** @internal */
+  private async cleanup(error?: Error) {
+    this[kClosed] = true;
+    const session = this[kSession];
+    try {
+      if (
+        !this[kKilled] &&
+        this[kId] &&
+        !this[kId].isZero() &&
+        this[kNamespace] &&
+        this[kServer] &&
+        !session.hasEnded
+      ) {
+        this[kKilled] = true;
+        await executeOperation(
+          this[kClient],
+          new KillCursorsOperation(this[kId], this[kNamespace], this[kServer], { session })
+        );
+      }
+    } catch (error) {
+      squashError(error);
+    } finally {
+      if (session?.owner === this) {
         await session.endSession({ error });
-        return;
       }
-
-      if (!session.inTransaction()) {
+      if (!session?.inTransaction()) {
         maybeClearPinnedConnection(session, { error });
       }
-    }
 
-    return;
+      this.emitClose();
+    }
   }
 
-  async function completeCleanup() {
-    if (session) {
-      if (session.owner === cursor) {
-        try {
-          await session.endSession({ error });
-        } finally {
-          cursor.emit(AbstractCursor.CLOSE);
-        }
-        return;
-      }
-
-      if (!session.inTransaction()) {
-        maybeClearPinnedConnection(session, { error });
+  /** @internal */
+  private hasEmittedClose = false;
+  /** @internal */
+  private emitClose() {
+    try {
+      if (!this.hasEmittedClose && (this[kDocuments].length === 0 || this[kClosed])) {
+        // @ts-expect-error: CursorEvents is generic so Parameters<CursorEvents["close"]> may not be assignable to `[]`. Not sure how to require extenders do not add parameters.
+        this.emit('close');
       }
+    } finally {
+      this.hasEmittedClose = true;
     }
-
-    cursor.emit(AbstractCursor.CLOSE);
-    return;
   }
 
-  cursor[kKilled] = true;
+  /** @internal */
+  private makeSafeTransform<TSchema>(transform: (doc: TSchema) => any): (doc: TSchema) => any {
+    return doc => {
+      try {
+        const result = transform(doc);
+        // eslint-disable-next-line no-restricted-syntax
+        if (result === null) {
+          const message =
+            'Cursor returned a `null` document, but the cursor is not exhausted.  Mapping documents to `null` is not supported in the cursor transform.';
+          throw new MongoAPIError(message);
+        }
 
-  if (session.hasEnded) {
-    return await completeCleanup();
+        return result;
+      } catch (error) {
+        // eslint-disable-next-line github/no-then
```

Comment:
(optional) it might make more sense to propagate errors from `transform` and instead try-catch the application of the transform each place we use it (which, I think is just one place? `next()`?)

It seems like a small code smell to wrap a synchronous function in another synchronous function that sometimes kicks off an async operation but doesn't await it, when we can remove the async portion of the transform wrapper by calling await cursor.close() in the caller, not the transform.

Reply:
Or there's no reason `makeSafeTransform` can't be async, since we await it everywhere we use it anyways.

Reply:
`transformDocument` now awaits the close! and I removed the makeTransform from stream because we didn't disallow null / try/catch previously